ARCHITECTURE.md

```diff
@@ -180,16 +180,80 @@ Issues with this approach:
+
+The user code would look something like:
+
+```js
+import Uppy, { Remote, Uploader } from 'uppy'
+import DriveBrowser from 'uppy-ui'
+
+const uppy = new Uppy(someOptions)
+uppy.use(Remote, { source: 'drive' })
+uppy.use(Uploader)
+
+uppy.render(DriveBrowser, document.getElementById('root'))
+```
+
+The DriveBrowser component might look like:
+
+```js
+import { Remote } from 'uppy'
+const googleDrive = new Remote({ source: 'drive' })
+const { actions } = googleDrive
+
+const DriveBrowser = (props) => {
+  return html`
+    <div>
+      <button onclick=${props.dispatch(actions.list('root'))}>Get list of files</button>
+      /* ... */
+    </div>
+  `
+}
+```
+
+Here is an example of what the plugin itself may look like: https://github.com/transloadit/uppy/blob/master/src/experimental/plugins/Remote.js
+
+Some things to note here:
+1. `Remote` is a generic plugin to interface with any remote provider, like Google Drive, Dropbox, Instagram, etc.
+
+2. Another one of the benefits to this approach is that we do away with all of the boilerplate methods that we currently have: `install`, `focus`, `update`, and `mount`.
+
+3. The `Remote` plugin example linked above extends `Remote` from a module I created called `uppy-base`. I created `uppy-base` while trying to figure out how to make Uppy work with React.  I took the bare functionality of Uppy that could be used universally (vanilla, React, Angular, anywhere) and put it into its own module.  The `uppy-base` plugins are stateless and not concerned with the UI.  Here is the `Remote` plugin from `uppy-base`: https://github.com/hedgerh/uppy-base/blob/master/src/plugins/Remote.js
+
+## React
+I'd like to give React users the freedom to manage their own Uppy state, whether they use internal state, or a state management lib like Redux.  The proposed solution of syncing the user's state to Uppy's internal state via event listeners, ie. `uppy.on('file-added', (file) => this.props.dispatch({ type: 'ADD_FILE', payload: file }))` is problematic because it violates the "single source of truth" principal, in that our state exists in two separate locations.  If we add a file, then roll back the add file action with time-travel debugging, the file will be removed from Redux's state, but still present in Uppy's state, causing de-synchronization between the two.
+
+The easiest solution I've found is to just use the `uppy-base` plugins directly in React, without Uppy's core.  Here is a comparison of core vs. no core: https://gist.github.com/hedgerh/9ad63f467ce816246044f3f9e83bd7e7
+
+For ease of use, we'd write a React container/wrapper component that abstracted away all of that from the user.  I've started on something like that here (bear in mind it's a bit messy):  https://github.com/hedgerh/uppy-react/blob/master/src/containers/UppyContainer.js
+
+and started on an example usage here: https://github.com/hedgerh/uppy-react/tree/master/examples/modal
+
```

Comment:
React proposal here seems cool! Downsides and critique are missing though, to me they are:

* Not clear how everything operates: each user comes up with his/her own method names, event names, file shapes.
* If there is no Core, what calculates total progress, watches for internet connection going online/offline, knows about capabilities? What adds metadata to files?
* Plugins behave differently (which might be fine if their base version has the same API), but seems like we’ll have Tus plugin for Vanilla Uppy and then another Tus for React Uppy.

Basically, the problem is things getting out of sync between two versions and current React version not taking into account things like metadata and file shaping (Core’s [addFile()](https://github.com/transloadit/uppy/blob/master/src/core/Core.js#L109-L157)). *Could be solved by having a shared Utils lib that does all those things and replaces Core in that sense*.

In [Core vs no-Core](https://gist.github.com/hedgerh/9ad63f467ce816246044f3f9e83bd7e7#file-uppyreduxcontainernocore-js) example you are effectively replacing Core with Uploader (which might be Tus or Multipart). That example probably shows our argument/discussion best — should we keep files in Core or manage them ourselves and pass to uploader plugin.

Also is Kevin’s proposal to hot-swap Core’s internal state with Redux state not possible? When you do `uppy.addFile(file)` and it automagically ends up in Redux state instead of Uppy’s internal.